desc/integrals/basis.py

```diff
@@ -58,6 +85,621 @@ def _in_epigraph_and(is_intersect, df_dy_sign, /):
+            Shape z.shape.
+            Isomorphic coordinates.
+
+        """
+        x_idx, y_val = jnp.divmod(z - self.domain[0], self.domain[-1] - self.domain[0])
+        x_idx = x_idx.astype(int)
+        y_val += self.domain[0]
+        return x_idx, y_val
+
+    def eval1d(self, z, cheb=None):
+        """Evaluate piecewise Chebyshev series at coordinates z.
+
+        Parameters
+        ----------
+        z : jnp.ndarray
+            Shape (..., *cheb.shape[:-2], z.shape[-1]).
+            Coordinates in [sef.domain[0], ∞).
+            The coordinates z ∈ ℝ are assumed isomorphic to (x, y) ∈ ℝ² where
+            ``z // domain`` yields the index into the proper Chebyshev series
+            along the second to last axis of ``cheb`` and ``z % domain`` is
+            the coordinate value on the domain of that Chebyshev series.
+        cheb : jnp.ndarray
+            Shape (..., M, N).
+            Chebyshev coefficients to use. If not given, uses ``self.cheb``.
+
+        Returns
+        -------
+        f : jnp.ndarray
+            Shape z.shape.
+            Chebyshev series evaluated at z.
+
+        """
+        cheb = _chebcast(setdefault(cheb, self.cheb), z)
+        N = cheb.shape[-1]
+        x_idx, y = self.isomorphism_to_C2(z)
+        y = bijection_to_disc(y, self.domain[0], self.domain[1])
+        # Chebyshev coefficients αₙ for f(z) = ∑ₙ₌₀ᴺ⁻¹ αₙ(x[z]) Tₙ(y[z])
+        # are held in cheb with shape (..., num cheb series, N).
+        cheb = jnp.take_along_axis(cheb, x_idx[..., jnp.newaxis], axis=-2)
+        f = idct_non_uniform(y, cheb, N)
+        assert f.shape == z.shape
+        return f
+
+    def intersect2d(self, k=0.0, *, eps=_eps):
+        """Coordinates yᵢ such that f(x, yᵢ) = k(x).
+
+        Parameters
+        ----------
+        k : jnp.ndarray
+            Shape must broadcast with (..., *cheb.shape[:-1]).
+            Specify to find solutions yᵢ to f(x, yᵢ) = k(x). Default 0.
+        eps : float
+            Absolute tolerance with which to consider value as zero.
+
+        Returns
+        -------
+        y : jnp.ndarray
+            Shape (..., *cheb.shape[:-1], N - 1).
+            Solutions yᵢ of f(x, yᵢ) = k(x), in ascending order.
+        is_intersect : jnp.ndarray
+            Shape y.shape.
+            Boolean array into ``y`` indicating whether element is an intersect.
+        df_dy_sign : jnp.ndarray
+            Shape y.shape.
+            Sign of ∂f/∂y (x, yᵢ).
+
+        """
+        c = _subtract_first(_chebcast(self.cheb, k), k)
+        # roots yᵢ of f(x, y) = ∑ₙ₌₀ᴺ⁻¹ αₙ(x) Tₙ(y) - k(x)
+        y = chebroots_vec(c)
+        assert y.shape == (*c.shape[:-1], self.N - 1)
+
+        # Intersects must satisfy y ∈ [-1, 1].
+        # Pick sentinel such that only distinct roots are considered intersects.
+        y = _filter_distinct(y, sentinel=-2.0, eps=eps)
+        is_intersect = (jnp.abs(y.imag) <= eps) & (jnp.abs(y.real) < 1.0)
```

Comment:
could evaluate in differentiable manner at y=+-1 by summing coeffcients as done in `stitch` function, but again probably not needed